File: diameter_of_binary_tree.ts
// 543. Diameter of Binary Tree
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

// DFS.
// Time: O(n), Space: O(h)
function diameterOfBinaryTree(root: TreeNode | null): number {
  let result = 0;

  function height(root: TreeNode | null): number {
    if (!root) return 0;

    const left = height(root.left);
    const right = height(root.right);
    result = Math.max(result, left + right);

    return Math.max(left, right) + 1;
  }

  height(root);

  return result;
}

// TODO: Iterative DFS

File: src/diameter_of_binary_tree.go
// 543. Diameter of Binary Tree
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
	result := 0

	var height func(node *TreeNode) int
	height = func(node *TreeNode) int {
		if node == nil {
			return 0
		}

		leftHeight := height(node.Left)
		rightHeight := height(node.Right)
		result = max(result, leftHeight+rightHeight)

		return max(leftHeight, rightHeight) + 1
	}

	height(root)

	return result
}
